### PR TITLE
Working Update to 2.0.7

### PR DESCRIPTION
**Pull request by eliasstepanik**

### CODE REVIEW

**Comment by Aareksio** [comment]
I am positive the hash you use is a hash of the factorio headless bundle, not the docker container. Docker build job failed.

**Comment by LsHallo** [comment]
I don't thinks this will work. update.sh needs to be adjusted.

**Comment by eliasstepanik** [comment]
> I don't thinks this will work. update.sh needs to be adjusted.

Worked for me.

**Comment by LsHallo** [comment]
Yeah but you are not github actions running update.sh and overwriting docker-compose.yml and buildinfo.json when you run.

**Comment by miragecentury** [comment]
```
stable_sha256=$(curl "https://factorio.com/download/sha256sums/" | grep "factorio_headless_x64_${stable_online_version}.tar.xz" | awk '{print $1}')
experimental_sha256=$(curl "https://factorio.com/download/sha256sums/" | grep "factorio_headless_x64_${experimental_online_version}.tar.xz" | awk '{print $1}')
```

Need to fix the filename

**Comment by LsHallo** [comment]
> Need to fix the filename

But only from 1.2.0 onward. That's the hard part im hung up on rn.

**Comment by miragecentury** [comment]
No need the script only take new version and from my understanding, they will never release any 1.... version.
The game (without the DLC) will be keep in the 2... core.

**Comment by HeartyShow** [comment]
Can't we do a simple condition then, considering they'll probably keep the new naming scheme from now on ? Or do we want to do some regex shanenigans to look for `headless` and `2.x.x` ?

**Comment by jonojr** [comment]
This greps using both styles of string, in case they go back and forth:
`curl "https://factorio.com/download/sha256sums/" | grep -E "factorio_headless_x64_${stable_online_version}.tar.xz|factorio-headless_linux_${stable_online_version}.tar.xz" | awk '{print $1}'`

**Comment by MrBasicX** [comment]
curl "https://factorio.com/download/sha256sums/" | grep -E "(factorio_headless_x64_|factorio-headless_linux_)${stable_online_version}.tar.xz" | awk '{print $1}'

**Comment by Fank** [comment]
Tell when I can merge

**Review by HeartyShow — COMMENTED** [review]
Nice code cleaning/commenting and fix at the same time

**Comment by eliasstepanik** [comment]
> Nice code cleaning/commenting and fix at the same time

Needs testing!

**Comment by HeartyShow** [comment]
Could you revert the commits of the hardcoded SHAs though

**Comment by eliasstepanik** [comment]
At the moment I am still working so could somebody test the update.sh.

**Comment by LsHallo** [comment]
> At the moment I am still working so could somebody test the update.sh.

Let it run for 5min. Does not seem to be exiting.
Gets to after sha256 sum retrieve step then hangs.

Output:
```
after fetch
after sha256
```

modified with debug output:
```
#!/bin/bash
set -e
SEMVER_REGEX="^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)$"

# Fetch online stable and experimental versions
stable_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.stable.headless' -r)
experimental_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.experimental.headless' -r)

echo "after fetch"

# Fetch SHA256 checksums
stable_sha256=$(curl -s "https://factorio.com/get-download/${stable_online_version}/headless/linux64" | grep "factorio_headless_x64_${stable_online_version}.tar.xz" | awk '{print $1}')
experimental_sha256=$(curl -s "https://factorio.com/get-download/${experimental_online_version}/headless/linux64" | grep "factorio_headless_x64_${experimental_online_version}.tar.xz" | awk '{print $1}')

echo "after sha256"

# Current versions in buildinfo.json
stable_current_version=$(jq -r 'with_entries(select(contains({value:{tags:["stable"]}}))) | keys | .[0]' buildinfo.json)
latest_current_version=$(jq -r 'with_entries(select(contains({value:{tags:["latest"]}}))) | keys | .[0]')

echo "from buildinfo"

# Output version information
echo "stable_online_version=${stable_online_version} experimental_online_version=${experimental_online_version}"
echo "stable_current_version=${stable_current_version} latest_current_version=${latest_current_version}"

echo "after version match"
```

File: update.sh [hunk 1]
```diff
@@ -2,22 +2,29 @@
...
+stable_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.stable.headless' -r)
+experimental_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.experimental.headless' -r)
+
+# Fetch SHA256 checksums
+stable_sha256=$(curl -s "https://factorio.com/get-download/${stable_online_version}/headless/linux64" | grep "factorio_headless_x64_${stable_online_version}.tar.xz" | awk '{print $1}')
+experimental_sha256=$(curl -s "https://factorio.com/get-download/${experimental_online_version}/headless/linux64" | grep "factorio_headless_x64_${experimental_online_version}.tar.xz" | awk '{print $1}')
+
+# Current versions in buildinfo.json
+stable_current_version=$(jq -r 'with_entries(select(contains({value:{tags:["stable"]}}))) | keys | .[0]' buildinfo.json)
+latest_current_version=$(jq -r 'with_entries(select(contains({value:{tags:["latest"]}}))) | keys | .[0]')
```
Comment (Fank) [inline update.sh:15]:
```suggestion
latest_current_version=$(jq -r 'with_entries(select(contains({value:{tags:["latest"]}}))) | keys | .[0]' buildinfo.json)
```

I guess

**Review by Fank — CHANGES_REQUESTED** [review]

File: update.sh (hunk 1 above)
Comment (LsHallo) [reply update.sh:15]: yep that does it

**Comment by jonojr** [comment]
> At the moment I am still working so could somebody test the update.sh.

I can confirm that it is not working, looking at fixes now

**Comment by LsHallo** [comment]
`yq: -i/--in-place can only be used with -y/-Y` now this happens

File: update.sh [hunk 2]
```diff
@@ -2,22 +2,29 @@
...
-experimental_sha256=$(curl "https://factorio.com/download/sha256sums/" | grep "factorio_headless_x64_${experimental_online_version}.tar.xz" | awk '{print $1}')
-stable_current_version=$(jq 'with_entries(select(contains({value:{tags:["stable"]}}))) | keys | .[0]' buildinfo.json -r)
-latest_current_version=$(jq 'with_entries(select(contains({value:{tags:["latest"]}}))) | keys | .[0]' buildinfo.json -r)
+# Fetch online stable and experimental versions
+stable_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.stable.headless' -r)
+experimental_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.experimental.headless' -r)
+
+# Fetch SHA256 checksums
+stable_sha256=$(curl -s "https://factorio.com/get-download/${stable_online_version}/headless/linux64" | grep "factorio_headless_x64_${stable_online_version}.tar.xz" | awk '{print $1}')
+experimental_sha256=$(curl -s "https://factorio.com/get-download/${experimental_online_version}/headless/linux64" | grep "factorio_headless_x64_${experimental_online_version}.tar.xz" | awk '{print $1}')
```
Comment (jonojr) [inline update.sh:11]: There are fetching the binaries, not the checksums

File: update.sh [hunk 3]
```diff
@@ -2,22 +2,29 @@
...
-stable_sha256=$(curl "https://factorio.com/download/sha256sums/" | grep "factorio_headless_x64_${stable_online_version}.tar.xz" | awk '{print $1}')
-experimental_sha256=$(curl "https://factorio.com/download/sha256sums/" | grep "factorio_headless_x64_${experimental_online_version}.tar.xz" | awk '{print $1}')
-stable_current_version=$(jq 'with_entries(select(contains({value:{tags:["stable"]}}))) | keys | .[0]' buildinfo.json -r)
-latest_current_version=$(jq 'with_entries(select(contains({value:{tags:["latest"]}}))) | keys | .[0]' buildinfo.json -r)
+# Fetch online stable and experimental versions
+stable_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.stable.headless' -r)
+experimental_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.experimental.headless' -r)
+
+# Fetch SHA256 checksums
+stable_sha256=$(curl -s "https://factorio.com/get-download/${stable_online_version}/headless/linux64" | grep "factorio_headless_x64_${stable_online_version}.tar.xz" | awk '{print $1}')
```
Comment (jonojr) [inline update.sh:10]:
```suggestion
stable_sha256=$(curl -s "https://factorio.com/download/sha256sums/" | grep -E"(factorio_headless_x64|factorio-headless_linux)_${stable_online_version}.tar.xz" | awk '{print $1}')
```

File: update.sh (hunk 2 above)
Comment (jonojr) [inline update.sh:11]:
```suggestion
stable_sha256=$(curl -s "https://factorio.com/download/sha256sums/" | grep -E"(factorio_headless_x64|factorio-headless_linux)_${experimental_online_version}.tar.xz" | awk '{print $1}')
```

File: update.sh [hunk 4]
```diff
@@ -2,22 +2,29 @@
...
-experimental_sha256=$(curl "https://factorio.com/download/sha256sums/" | grep "factorio_headless_x64_${experimental_online_version}.tar.xz" | awk '{print $1}')
-stable_current_version=$(jq 'with_entries(select(contains({value:{tags:["stable"]}}))) | keys | .[0]' buildinfo.json -r)
-latest_current_version=$(jq 'with_entries(select(contains({value:{tags:["latest"]}}))) | keys | .[0]' buildinfo.json -r)
+# Fetch online stable and experimental versions
+stable_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.stable.headless' -r)
+experimental_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.experimental.headless' -r)
+
+# Fetch the binaries
+stable_sha256=$(curl -s "https://factorio.com/get-download/${stable_online_version}/headless/linux64" | grep "factorio_headless_x64_${stable_online_version}.tar.xz" | awk '{print $1}')
+stable_sha256=$(curl -s "https://factorio.com/download/sha256sums/" | grep -E"(factorio_headless_x64|factorio-headless_linux)_${experimental_online_version}.tar.xz" | awk '{print $1}')
```
Comment (jonojr) [inline update.sh:11]: This is the experimental version SHA, we are using the `experimental_online_version`
```suggestion
experimental_sha256=$(curl -s "https://factorio.com/download/sha256sums/" | grep -E"(factorio_headless_x64|factorio-headless_linux)_${experimental_online_version}.tar.xz" | awk '{print $1}')
```

Comment (Sculas) [inline update.sh:11]: Previous review accidentally removed the experimental SHA256 checksum:
```suggestion
experimental_sha256=$(curl -s "https://factorio.com/download/sha256sums/" | grep -E"(factorio_headless_x64|factorio-headless_linux)_${experimental_online_version}.tar.xz" | awk '{print $1}')
```

**Review by jonojr — CHANGES_REQUESTED** [review]
Need to fix both of the commands getting the hashes and fix the name

File: update.sh (hunk 4 above)
Comment (jonojr) [reply update.sh:11]: My bad on that 😅

File: update.sh (hunk 3 above)
Comment (jonojr) [reply update.sh:10]: We also need this change @eliasstepanik as both the stable & experimental need updating

**Comment by eliasstepanik** [comment]
Sadly I don't have anymore time today. I hope you can resolve the Problems.

File: update.sh [hunk 5]
```diff
@@ -2,22 +2,29 @@
...
-experimental_online_version=$(curl 'https://factorio.com/api/latest-releases' | jq '.experimental.headless' -r)
-stable_sha256=$(curl "https://factorio.com/download/sha256sums/" | grep "factorio_headless_x64_${stable_online_version}.tar.xz" | awk '{print $1}')
-experimental_sha256=$(curl "https://factorio.com/download/sha256sums/" | grep "factorio_headless_x64_${experimental_online_version}.tar.xz" | awk '{print $1}')
-stable_current_version=$(jq 'with_entries(select(contains({value:{tags:["stable"]}}))) | keys | .[0]' buildinfo.json -r)
-latest_current_version=$(jq 'with_entries(select(contains({value:{tags:["latest"]}}))) | keys | .[0]' buildinfo.json -r)
+# Fetch online stable and experimental versions
+stable_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.stable.headless' -r)
+experimental_online_version=$(curl -s 'https://factorio.com/api/latest-releases' | jq '.experimental.headless' -r)
+
+# Fetch the binaries
```
Comment (jonojr) [inline update.sh:9]:
```suggestion
# Fetch the checksums
```

File: update.sh [hunk 6]
```diff
@@ -2,22 +2,29 @@
...
+stable_sha256=$(curl -s "https://factorio.com/get-download/${stable_online_version}/headless/linux64" | grep "factorio_headless_x64_${stable_online_version}.tar.xz" | awk '{print $1}')
```
Comment (jonojr) [inline update.sh:10]:
```suggestion
stable_sha256=$(curl -s "https://factorio.com/download/sha256sums/" | grep -E"(factorio_headless_x64|factorio-headless_linux)_${stable_online_version}.tar.xz" | awk '{print $1}')
```

**Comment by miragecentury** [comment]
https://github.com/factoriotools/factorio-docker/pull/505 ? seem the good fixes ?

**Comment by jonojr** [comment]
Thinking we might need to move to: https://github.com/factoriotools/factorio-docker/pull/505

**Comment by Fank** [comment]
I merged #505 for a quick fix, but I would like to see this change as well, because of the cleanup.
There are conflicts now, but these should be small.

**Comment by miragecentury** [comment]
> I merged #505 for a quick fix, but I would like to see this change as well, because of the cleanup. There are conflicts now, but these should be small.

Yes but it will be better to just have an MR with the cleanup. Here, we clean also the .json file.

**Pull request closed without merging**